### PR TITLE
Source SFTP-Bulk: Support custom CSV separators

### LINKED ISSUE #15844: [SFTP] Choose CSV separator
## Tell us about the problem you're trying to solve

When adding a SFTP source fetching CSV files, we cannot change the **default separator** that is `,`. In fact, when creating a connection containing a SFTP source, the *schema fetching phase* fails with this error in the UI:

image

When digging in the logs, I can find this line: 
```java
2022-08-22 12:30:50 [32mINFO[m i.a.w.i.DefaultAirbyteStreamFactory(lambda$create$0):61 - 2022-08-22 12:30:50 [1;31mERROR[m i.a.i.s.s.SftpCommand(tryGetFirstNode):109 - Exception occurred while trying to parse file XXX.csv : 
2022-08-22 12:30:50 [32mINFO[m i.a.w.i.DefaultAirbyteStreamFactory(lambda$create$0):61 - com.fasterxml.jackson.core.JsonParseException: Unexpected character (';' (code 59)): Expected column separator character (',' (code 44)) or end-of-line
```

So, Airbyte **cannot parse** the CSV files on the SFTP server because its separator is `;`.

## Describe the solution you’d like

It would be a nice feature to be able to **select the separator** when creating a SFTP source fetching CSV files.

## Describe the alternative you’ve considered or used

I cannot use any workaround because I am not the producer of these CSV files, so I cannot change their structure.

## Are you willing to submit a PR?

No, I don't have the knowledge to fix this.

### PR DESCRIPTION
**Pull request by anamargaridarl**
## What
Allow different types of separators in CSV files 

Closes https://github.com/airbytehq/airbyte/issues/15844

## How
Use Sniffer to define the separator when reading the csv file.

## Recommended reading order
1. sftp-bulk/client.py

## 🚨 User Impact 🚨
Low

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [x] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

## Tests

<details><summary><strong>Unit</strong></summary>

> Task :airbyte-integrations:connectors:source-sftp-bulk:unitTest

> Name                           Stmts   Miss  Cover
> --------------------------------------------------
> source_sftp_bulk/__init__.py       2      0   100%
> source_sftp_bulk/source.py        72     39    46%
> source_sftp_bulk/streams.py       43     26    40%
> source_sftp_bulk/client.py       138    101    27%
> --------------------------------------------------
> TOTAL                            255    166    35%

</details>

<details><summary><strong>Integration</strong></summary>

</details>

<details><summary><strong>Acceptance</strong></summary>

	 ---------- coverage: platform darwin, python 3.9.11-final-0 ----------
	 Name                                                 Stmts   Miss  Cover   Missing
	 ----------------------------------------------------------------------------------
	 source_acceptance_test/base.py                          12      4    67%   16-19
	 source_acceptance_test/config.py                       140      5    96%   87, 93, 238, 242-243
	 source_acceptance_test/conftest.py                     208     92    56%   36, 42-44, 49, 54, 77, 83, 89-91, 110, 115-117, 123-125, 131-132, 137-138, 143, 149, 158-167, 173-178, 193, 217, 248, 254, 262-267, 275-280, 288-301, 306-312, 319-330, 337-353
	 source_acceptance_test/plugin.py                        69     25    64%   22-23, 31, 36, 120-140, 144-148
	 source_acceptance_test/tests/test_core.py              402    115    71%   53, 58, 93-104, 109-116, 120-121, 125-126, 308, 346-363, 376-387, 391-396, 402, 435-440, 478-485, 528-530, 533, 598-606, 618-621, 626, 682-683, 689, 692, 728-738, 751-776
	 source_acceptance_test/tests/test_incremental.py       158     14    91%   52-59, 64-77, 240
	 source_acceptance_test/utils/asserts.py                 39      2    95%   62-63
	 source_acceptance_test/utils/common.py                  94      8    91%   32-38, 72, 75
	 source_acceptance_test/utils/compare.py                 62     23    63%   21-51, 68, 97-99
	 source_acceptance_test/utils/connector_runner.py       133     33    75%   24-27, 46-47, 50-54, 57-58, 73-75, 78-80, 83-85, 88-90, 93-95, 124-125, 159-161, 208
	 source_acceptance_test/utils/json_schema_helper.py     107     13    88%   30-31, 38, 41, 65-68, 96, 120, 192-194
	 ----------------------------------------------------------------------------------
	 TOTAL                                                 1603    334    79%
	 
	 5 files skipped due to complete coverage.
	 
	 Required test coverage of 74.0% reached. Total coverage: 79.16%
	 
	 Results (131.68s):
	      436 passed

------------------------

test_core.py ........................s..                                 [ 87%]
test_full_refresh.py .                                                   [ 90%]
test_incremental.py ...                                                  [100%]	     

</details>

### CODE REVIEW

**Comment by anamargaridarl** [comment]
I'm having issues running the tests related to the temporary sftp server to run the tests against. I've seen issues with this in the first PR. Could someone help me with this?

>  docker.errors.APIError: 409 Client Error for http+docker://localhost/v1.41/containers/create?name=mysftpssh: Conflict ("Conflict. The container name "/mysftpssh" is already in use by container "247bfd28793b3c43b711cf7c4a6de0fd5ae380b33517a439ed9f1c5d2b6641cd". You have to remove (or rename) that container to be able to reuse that name.")

**Comment by sathiscode** [comment]
Try removing your stopped containers and run again. That should solve the issue.

File: airbyte-integrations/connectors/source-sftp-bulk/source_sftp_bulk/client.py [hunk 1]
```diff
@@ -168,15 +169,22 @@ def get_files(self, prefix, search_pattern=None, modified_since=None, most_recen
...
         return sorted_files
 
+    def peek_line(self, f):
+        pos = f.tell()
+        line = f.readline()
+        f.seek(pos)
+        return line
+
     @backoff.on_exception(backoff.expo, (socket.timeout), max_tries=5, factor=2)
     def fetch_file(self, fn: Mapping[str, Any], file_type="csv") -> pd.DataFrame:
```
Comment (sathiscode) [inline airbyte-integrations/connectors/source-sftp-bulk/source_sftp_bulk/client.py:179]: Is it possible to add a parameter for separator, so user will have more control over the separator. If there is no separator passed, then Sniffer can be used.

**Comment by anamargaridarl** [comment]
> Try removing your stopped container (one with the name mysftpssh) and run again. Most likely that will solve the issue.

Yes, thank you, I had to restart the docker to work. But now I get this error even if I run it in master. So it's probably something wrongly configured from my side?

>  FileNotFoundError: [Errno 2] No such file or directory: '/test_input/secrets/config.json'

**Comment by sathiscode** [comment]
sftp credentials is expected in the config.json. Please check the integration_tests/valid_config.json for a sample of the config.json file. Also, please read README.md file in source-sftp-bulk directory.

**Comment by marcosmarxm** [comment]
Hello :wave:, first thank you for this amazing contribution. 

We really appreciate the effort you've made to improve the project.
We ask you patience for the code review. Last month our team was focused on Hacktoberfest event and that probably left some PR without the proper feedback. And this week, due to the Thanksgiving US Holiday, most our team is out of office with their families. Another important piece of information why code won't be merge this week is: as a safety measure the core team has decided to freeze merging code to main branch to keep the release stable. Next week we'll return to you with the proper code review and update the status of your contribution.

If you have any questions feel free to send me a message in Slack!
Thanks!

**Comment by anamargaridarl** [comment]
I will try to address the changes this week. Thanks for the feedback!

File: airbyte-integrations/connectors/source-sftp-bulk/source_sftp_bulk/client.py [hunk 2]
```diff
@@ -168,15 +169,25 @@ def get_files(self, prefix, search_pattern=None, modified_since=None, most_recen
...
 
+    def peek_line(self, f):
+        pos = f.tell()
+        line = f.readline()
+        f.seek(pos)
+        return line
+
     @backoff.on_exception(backoff.expo, (socket.timeout), max_tries=5, factor=2)
-    def fetch_file(self, fn: Mapping[str, Any], file_type="csv") -> pd.DataFrame:
+    def fetch_file(self, fn: Mapping[str, Any], separator, file_type="csv") -> pd.DataFrame:
```
Comment (sathiscode) [inline airbyte-integrations/connectors/source-sftp-bulk/source_sftp_bulk/client.py:179]: Can you please add default value for the separator?
#Option 1 - Add "," as default separator value (as default file type is csv) 
#Option 2 - Add None as default separator value (but the existing pipelines which uses default values will incur Sniffer processing overhead.
So my suggestion is to use option 1 and add comments as passing None as separator will try to identify the separator using Sniffer.

File: airbyte-integrations/connectors/source-sftp-bulk/source_sftp_bulk/client.py [hunk 3]
```diff
@@ -196,10 +207,10 @@ def fetch_file(self, fn: Mapping[str, Any], file_type="csv") -> pd.DataFrame:
 
             raise Exception("Unable to read file: %s" % e) from e
 
-    def fetch_files(self, files, file_type="csv") -> Tuple[datetime, Dict[str, Any]]:
+    def fetch_files(self, files, separator, file_type="csv") -> Tuple[datetime, Dict[str, Any]]:
```
Comment (sathiscode) [inline airbyte-integrations/connectors/source-sftp-bulk/source_sftp_bulk/client.py:210]: Can you please add default value for the separator?
#Option 1 - Add "," as default separator value (as default file type is csv) 
#Option 2 - Add None as default separator value (but the existing pipelines which uses default values will incur Sniffer processing overhead.
So my suggestion is to use option 1 and add comments as passing None as separator will try to identify the separator using Sniffer.

File: airbyte-integrations/connectors/source-sftp-bulk/source_sftp_bulk/client.py (hunk 2 above)
Comment (sathiscode) [reply airbyte-integrations/connectors/source-sftp-bulk/source_sftp_bulk/client.py:179]: @anamargaridarl  the method definition is not updated. Can you please recheck it?

Comment (sathiscode) [reply airbyte-integrations/connectors/source-sftp-bulk/source_sftp_bulk/client.py:179]: `def fetch_file(self, fn: Mapping[str, Any], separator=",", file_type="csv")`

**Comment by marcosmarxm** [comment]
Hello :wave::skin-tone-2: and thank you for your contribution!

Airbyte has instituted a code freeze between 19 and 30 December, to make sure there are no disruptions during the holidays. 
Because of this, reviewing and merging your contribution may take longer than usual. 
We apologize for the delay, but we want everyone to have a quiet and happy holiday.

If you have any questions or need further clarification, please don't hesitate to ping via Slack.

**Comment by anamargaridarl** [comment]
Hi! Could I have someone check again in this PR? Thank you!

File: airbyte-integrations/connectors/source-sftp-bulk/source_sftp_bulk/spec.json [hunk 4]
```diff
@@ -98,6 +98,14 @@
...
         "description": "The date from which you'd like to replicate data for all incremental streams, in the format YYYY-MM-DDT00:00:00Z. All data generated after this date will be replicated.",
         "order": 10
+      },
+      "separator": {
+        "title": "CSV Separator (Optional)",
+        "description": "The separator used in the cvs files. Define None if you want to use the Sniffer functionality",
+        "type": "string",
+        "default": ",",
+        "examples": [","],
+        "order": 1
```
Comment (sh4sh) [inline airbyte-integrations/connectors/source-sftp-bulk/source_sftp_bulk/spec.json:108]: `order` should be unique, it determines the order the config options will display. To set it as the last item displayed in the connector settings:

```suggestion
        "order": 11
```

Alternatively, you can adjust the `order` values in this file so it is further up.

File: airbyte-integrations/connectors/source-sftp-bulk/integration_tests/files/csv/test_2.csv [hunk 5]
```diff
@@ -0,0 +1,3 @@
+string_col;int_col
+"hello";1
+"foo";2
```
Comment (sh4sh) [inline airbyte-integrations/connectors/source-sftp-bulk/integration_tests/files/csv/test_2.csv:3]:
```suggestion
"foo";2

```

File: airbyte-integrations/connectors/source-sftp-bulk/source_sftp_bulk/spec.json [hunk 6]
```diff
@@ -98,6 +98,14 @@
         "examples": ["2017-01-25T00:00:00Z"],
         "description": "The date from which you'd like to replicate data for all incremental streams, in the format YYYY-MM-DDT00:00:00Z. All data generated after this date will be replicated.",
         "order": 10
+      },
+      "separator": {
+        "title": "CSV Separator (Optional)",
+        "description": "The separator used in the cvs files. Define None if you want to use the Sniffer functionality",
```
Comment (sh4sh) [inline airbyte-integrations/connectors/source-sftp-bulk/source_sftp_bulk/spec.json:104]:
```suggestion
        "description": "The separator used in the CSV files. Define None if you want to use the Sniffer functionality",
```
I think this is a typo 😛

**Comment by sh4sh** [comment]
The failure is due to needing to fix the `order` property in spec.yaml, once that is resolved tests will be passing 😁

**Review by sh4sh — APPROVED** [review]

**Comment by sh4sh** [comment]
Merge broke master I think because I ran /publish on the testing PR. We can merge this from there

**Comment by marcosmarxm** [comment]
@sh4sh why you didn't publish the PR here?

**Pull request merged**